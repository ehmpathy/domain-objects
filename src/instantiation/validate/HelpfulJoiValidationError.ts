// only importing types  > dev dep
import type { ValidationError } from 'joi';

interface HelpfulJoiValidationErrorDetail {
  message: string;
  path: string;
  type: string;
}
export class HelpfulJoiValidationError extends Error {
  public details: HelpfulJoiValidationErrorDetail[];

  public props: any;

  public domainObject: string;

  constructor({
    error,
    props,
    domainObject,
  }: {
    error: ValidationError;
    props: any;
    domainObject: string;
  }) {
    const details = error.details.map((detail) => ({
      message: detail.message,
      path: detail.path.join('.'),
      type: detail.type,
    }));

    const message = `
Errors on ${
      Object.keys(details).length
    } properties were found while validating properties for domain object ${domainObject}.:
${JSON.stringify(details, null, 2)}

Props Provided:
${JSON.stringify(props, null, 2)}
    `.trim();
    super(message);

    this.details = details;
    this.props = props;
    this.domainObject = domainObject;
  }
}
