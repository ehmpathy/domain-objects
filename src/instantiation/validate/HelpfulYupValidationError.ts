import { ValidationError } from 'yup';

export class HelpfulYupValidationError extends Error {
  public details: string[];

  public props: any;

  public domainObject: string;

  constructor({ error, props, domainObject }: { error: ValidationError; props: any; domainObject: string }) {
    const message = `
Errors were found while validating properties for domain object ${domainObject}.:
${JSON.stringify(error.errors, null, 2)}

Props Provided:
${JSON.stringify(props, null, 2)}
    `.trim();
    super(message);

    this.details = error.errors;
    this.props = props;
    this.domainObject = domainObject;
  }
}
