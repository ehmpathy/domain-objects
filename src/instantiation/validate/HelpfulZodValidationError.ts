// only importing types  > dev dep
import type { ZodError, ZodIssue } from 'zod';

export class HelpfulZodValidationError extends Error {
  public details: ZodIssue[];

  public props: any;

  public domainObject: string;

  constructor({
    error,
    props,
    domainObject,
  }: {
    error: ZodError;
    props: any;
    domainObject: string;
  }) {
    const message = `
Errors were found while validating properties for domain object ${domainObject}.:
${JSON.stringify(error.errors, null, 2)}

Props Provided:
${JSON.stringify(props, null, 2)}
    `.trim();
    super(message);

    this.details = error.errors;
    this.props = props;
    this.domainObject = domainObject;
  }
}
