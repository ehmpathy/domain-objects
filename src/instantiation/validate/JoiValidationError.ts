import Joi from 'joi';

interface JoiValidationErrorDetail {
  message: string;
  path: string;
  type: string;
}
export class JoiValidationError extends Error {
  public details: JoiValidationErrorDetail[];

  public props: any;

  public domainObject: string;

  constructor({ error, props, domainObject }: { error: Joi.ValidationError; props: any; domainObject: string }) {
    const details = error.details.map((detail) => ({ message: detail.message, path: detail.path.join('.'), type: detail.type }));

    const message = `
Errors on ${Object.keys(details).length} properties were found while validating properties for domain object ${domainObject}.:
${JSON.stringify(details, null, 2)}

Props Provided:
${JSON.stringify(props, null, 2)}
    `.trim();
    super(message);

    this.details = details;
    this.props = props;
    this.domainObject = domainObject;
  }
}
