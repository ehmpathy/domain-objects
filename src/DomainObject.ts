import { Schema } from 'joi';
import { validate } from './validate/validate';

/**
 * Domain Object
 *
 * Responsibilities:
 * - optionally validate the properties that are passed into the constructor against the schema at runtime, if schema is supplied
 * - assign all properties that are passed into constructor to self, after optional runtime validation
 */
export class DomainObject<T> {
  constructor(props: T) {
    // 1. validate with the schema if provided
    const { schema } = this.constructor as typeof DomainObject; // `this.constructor` does not get typed to DomainObject automatically by ts; https://stackoverflow.com/questions/33387318/access-to-static-properties-via-this-constructor-in-typescript
    if (schema) validate({ props, schema, domainObjectName: this.constructor.name });

    // 2. assign all properties to self if passed validation
    Object.assign(this, props);
  }

  /**
   * DomainObject.schema
   *
   * When set, will be used to validate the properties passed into the constructor at runtime (i.e., during instantiation)
   */
  public static schema?: Schema;
}
