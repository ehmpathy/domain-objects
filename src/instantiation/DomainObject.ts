import { withImmute, WithImmute } from '../manipulation/immute/withImmute';
import { DomainObjectShape } from '../reference/DomainReferenceable';
import { hydrateNestedDomainObjects } from './hydrate/hydrateNestedDomainObjects';
import { SchemaOptions, validate } from './validate/validate';

export interface DomainObjectInstantiationOptions {
  /**
   * allow callers to skip certain aspects of instantiation
   *
   * e.g., for performance optimizations in specific usecases
   */
  skip?: {
    /**
     * allow callers to skip schema validation
     *
     * usecase examples
     * - deserialize: schema validation increases deserialization time dramatically (10-100x) (especially if using Joi, that thing is slow!)
     */
    schema?: boolean;
  };
}

/**
 * Domain Object
 *
 * Responsibilities:
 * - optionally validate the properties that are passed into the constructor against the schema at runtime, if schema is supplied
 * - assign all properties that are passed into constructor to self, after optional runtime validation
 */
export class DomainObject<T extends DomainObjectShape> {
  constructor(props: T, options?: DomainObjectInstantiationOptions) {
    // 1. validate with the schema if provided
    const { schema } = this.constructor as typeof DomainObject; // `this.constructor` does not get typed to DomainObject automatically by ts; https://stackoverflow.com/questions/33387318/access-to-static-properties-via-this-constructor-in-typescript
    if (schema && !(options?.skip?.schema === true))
      validate({ props, schema, domainObjectName: this.constructor.name });

    // 2. hydrate any nested props present; just overwrite the orig props for each "nested" key
    const nested = ((this.constructor as typeof DomainObject).nested ??
      {}) as Record<string, typeof DomainObject>;
    const hydratedProps = hydrateNestedDomainObjects({
      domainObjectName: this.constructor.name,
      props,
      nested,
    });

    // 3. assign all properties to self if passed validation
    Object.assign(this, hydratedProps);
  }

  /**
   * DomainObject.alias
   *
   * When set, declares the alias via which domain objects of this class are known to be referred to.
   *
   * For example
   * - PlantPot.alias = 'pot'
   * - Bird.alias = { singular: 'bird', plural: 'flock' }
   *
   * Relevance
   * - formalizes that we can expect to see the dobj referred to by this alias
   * - enables code generators to use this colloquial alias in contracts for better devex
   */
  public static alias?: string | { plural?: string; singular?: string };

  /**
   * DomainObject.schema
   *
   * When set, will be used to validate the properties passed into the constructor at runtime (i.e., during instantiation)
   *
   * Supports [`Zod`](https://github.com/colinhacks/zod), [`Joi`](https://github.com/sideway/joi), and [`Yup`](https://github.com/jquense/yup)
   */
  public static schema?: SchemaOptions<any>; // todo: work around typescript's "static members cant reference class type parameters"

  /**
   * DomainObject.nested
   *
   * When set, will be used to hydrate nested DomainObjects. This is especially useful when instantiating nested domain objects from data stores (e.g., apis or databases). This is _required_ in order to safely use the `serialize` and `getUniqueIdentifier` functions - as those functions check static properties of domain objects to function correctly.
   *
   * For example:
   * ```ts
   * // define the plant pot
   * interface PlantPot {
   *   diameterInInches: number;
   * }
   * class PlantPot extends DomainObject<PlantPot> implements PlantPot {}
   *
   * // define the plant owner
   * interface PlantOwner {
   *   diameterInInches: number;
   * }
   * class PlantOwner extends DomainObject<PlantOwner> implements PlantOwner {}
   *
   * // define the plant
   * interface Plant {
   *   pot: PlantPot;
   *   owners: PlantOwner[];
   *   lastWatered: string;
   * }
   * class Plant extends DomainObject<Plant> implements Plant {
   *   public static nested = { pot: PlantPot, owners: PlantOwner }
   * }
   *
   * // and show that it hydrates the plant pot
   * const plant = new Plant({ pot: { diameterInInches: 7 }, owners: [{ name: 'bob' }], lastWatered: 'monday' }); // notice how the param of `pot` we're inputting is _not_ an instance of `PlantPot`, and `owners` are not instances of `PlantOwner`
   * expect(plant.pot).toBeInstanceOf(PlantPot); // now succeeds, as during instantiation we hydrated `pot` into an instance of `PlantPot`, due to `Plant.nested.pot` being set
   * plant.owners.forEach((owner) => expect(owner).toBeInstance(PlantOwner)); // also succeeds, since during instantiation we hydrated each `owner` into an instance of `PlantOwner`, due to `Plant.nested.owners` being set
   * ```
   */
  public static nested?: Record<string, DomainObject<any>>; // TODO: find a way to make this Record<keyof string>

  /**
   * DomainObject.metadata
   *
   * When set, customizes the keys that are considered as metadata of this domain object.
   *
   * Context,
   * - domain objects are often persisted inside of storage mechanisms that assign metadata to them, such as ids or timestamps
   * - metadata simply adds information _about_ the object, without contributing to _defining_ the object
   *
   * Relevance,
   * - metadata properties do not contribute to the unique key of a DomainLiteral
   * - metadata properties can be easily stripped from an object by using the `omitMetadataValues` method
   *
   * By default,
   * - `id`, `createdAt`, `updatedAt`, and `effectiveAt` are considered metadata keys
   * - `uuid` is also considered a metadata key, if it is not included in the unique key of the DomainEntity or DomainEvent
   */
  public static metadata: readonly string[];

  /**
   * .what = an interface via which to construct instances w/ immute operations
   * .why =
   *   - immute operations such as .clone produce more maintainable code by preventing unexpected mutations
   *   - these immute operations provide a safe pit of success for common operations
   * .note =
   *   - you can add withImmute to any dobj yourself, even if it wasn't built via this .build procedure
   *   - you can override the .build to add your own domain's getters, too
   */
  static build<TInstance extends DomainObjectShape>(
    this: new (props: TInstance) => TInstance,
    props: ConstructorParameters<typeof this>[0],
  ): WithImmute<TInstance> {
    const instance = new this(props);
    return withImmute(instance);
  }
}
