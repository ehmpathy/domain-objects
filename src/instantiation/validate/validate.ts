import type { Schema as JoiSchema } from 'joi';
import type { ValidationError, Schema as YupSchema } from 'yup';

// only importing types -> dev dep
import { HelpfulJoiValidationError } from './HelpfulJoiValidationError';
import { HelpfulYupValidationError } from './HelpfulYupValidationError';

export type SchemaOptions = YupSchema<any> | JoiSchema;

const isJoiSchema = (schema: SchemaOptions): schema is JoiSchema => {
  if ((schema as JoiSchema).$) return true; // joi schemas have `$`, yup does not
  return false;
};

const isYupSchema = (schema: SchemaOptions): schema is YupSchema<any> => {
  // for now, since we only support two options, if its not a joi schema, it must be a yup schema
  return !isJoiSchema(schema);
};

export const validate = ({
  domainObjectName,
  schema,
  props,
}: {
  domainObjectName: string;
  schema: SchemaOptions;
  props: any;
}): void => {
  if (isJoiSchema(schema)) {
    const result = schema.validate(props);
    if (result.error)
      throw new HelpfulJoiValidationError({
        domainObject: domainObjectName,
        error: result.error,
        props,
      });
  }
  if (isYupSchema(schema)) {
    try {
      schema.validateSync(props);
    } catch (error) {
      if (!(error instanceof Error)) throw error;
      if (error.constructor.name === 'ValidationError')
        throw new HelpfulYupValidationError({
          domainObject: domainObjectName,
          error: error as ValidationError,
          props,
        }); // if we got a yup validation error, make it more helpful
      throw error; // otherwise throw the error we got
    }
  }
};
