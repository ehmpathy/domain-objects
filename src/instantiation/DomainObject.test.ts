import Joi from 'joi';
import { v4 as uuid } from 'uuid';
import * as yup from 'yup';
import { z } from 'zod';

import { DomainObject } from './DomainObject';
import { HelpfulJoiValidationError } from './validate/HelpfulJoiValidationError';
import { HelpfulYupValidationError } from './validate/HelpfulYupValidationError';
import { HelpfulZodValidationError } from './validate/HelpfulZodValidationError';

describe('DomainObject', () => {
  describe('domain modeling use cases', () => {
    it('should be able to represent a literal', () => {
      interface ChatMessage {
        userUuid: string;
        conversationUuid: string;
        message: string;
      }
      class ChatMessage
        extends DomainObject<ChatMessage>
        implements ChatMessage {}
      const message = new ChatMessage({
        userUuid: uuid(),
        conversationUuid: uuid(),
        message: 'Hello, World!',
      });
      expect(message).toBeInstanceOf(ChatMessage); // sanity check
    });
    it('should be able to represent an entity', () => {
      interface RocketShip {
        serialNumber: string;
        fuelQuantity: number;
        passengers: number;
      }
      class RocketShip extends DomainObject<RocketShip> implements RocketShip {}
      const ship = new RocketShip({
        serialNumber: uuid(),
        fuelQuantity: 9001,
        passengers: 21,
      });
      expect(ship).toBeInstanceOf(RocketShip); // sanity check
    });
  });

  describe('instantiation', () => {
    interface ChatMessage {
      userUuid: string;
      conversationUuid: string;
      message: string;
    }
    class ChatMessage
      extends DomainObject<ChatMessage>
      implements ChatMessage {}
    it('should assign all properties in the constructor to the instance', () => {
      const message = new ChatMessage({
        userUuid: '__USER_UUID__',
        conversationUuid: '__CONVO_UUID__',
        message: 'Hello, World!',
      });
      expect(message.userUuid).toEqual('__USER_UUID__');
      expect(message.conversationUuid).toEqual('__CONVO_UUID__');
      expect(message.message).toEqual('Hello, World!');
    });
    it('should be able to spread into itself', () => {
      const message = new ChatMessage({
        userUuid: '__USER_UUID__',
        conversationUuid: '__CONVO_UUID__',
        message: 'Hello, World!',
      });
      const updatedMessage = new ChatMessage({
        ...message,
        message: `Hello, World!\n Edit: You're great!`,
      });
      expect(updatedMessage.userUuid).toEqual('__USER_UUID__');
      expect(updatedMessage.conversationUuid).toEqual('__CONVO_UUID__');
      expect(updatedMessage.message).toEqual(
        `Hello, World!\n Edit: You're great!`,
      );
    });
  });

  describe('validation', () => {
    describe('Joi schema', () => {
      interface RocketShip {
        serialNumber: string;
        fuelQuantity: number;
        passengers: number;
      }
      const schema = Joi.object().keys({
        serialNumber: Joi.string().uuid().required(),
        fuelQuantity: Joi.number().required(),
        passengers: Joi.number().max(42).required(),
      });
      class RocketShip extends DomainObject<RocketShip> implements RocketShip {
        public static schema = schema;
      }
      it('should not throw error if when valid', () => {
        const ship = new RocketShip({
          serialNumber: uuid(),
          fuelQuantity: 9001,
          passengers: 21,
        });
        expect(ship).toBeInstanceOf(RocketShip); // sanity check
      });
      it('should throw a helpful error when does not pass joi schema', () => {
        try {
          // eslint-disable-next-line no-new
          new RocketShip({
            serialNumber: '__SOME_UUID__',
            fuelQuantity: 9001,
            passengers: 50,
          });
          throw new Error('should not reach here');
        } catch (error) {
          if (!(error instanceof Error)) throw error;
          expect(error).toBeInstanceOf(HelpfulJoiValidationError);
          expect(error.message).toMatchSnapshot();
        }
      });
    });
    describe('Yup schema', () => {
      interface RocketShip {
        serialNumber: string;
        fuelQuantity: number;
        passengers: number;
      }
      const schema = yup.object({
        serialNumber: yup.string().required(),
        fuelQuantity: yup.number().required(),
        passengers: yup.number().max(42).required(),
      });
      class RocketShip extends DomainObject<RocketShip> implements RocketShip {
        public static schema = schema;
      }
      it('should not throw error if when valid', () => {
        const ship = new RocketShip({
          serialNumber: uuid(),
          fuelQuantity: 9001,
          passengers: 21,
        });
        expect(ship).toBeInstanceOf(RocketShip); // sanity check
      });
      it('should throw a helpful error when does not pass schema', () => {
        try {
          // eslint-disable-next-line no-new
          new RocketShip({
            serialNumber: '__SOME_UUID__',
            fuelQuantity: 9001,
            passengers: 50,
          });
          throw new Error('should not reach here');
        } catch (error) {
          if (!(error instanceof Error)) throw error;
          expect(error).toBeInstanceOf(HelpfulYupValidationError);
          expect(error.message).toMatchSnapshot();
        }
      });
    });

    describe('Zod schema', () => {
      interface RocketShip {
        serialNumber: string;
        fuelQuantity: number;
        passengers: number;
      }
      const schema = z.object({
        serialNumber: z.string(),
        fuelQuantity: z.number(),
        passengers: z.number().max(42),
      });
      class RocketShip extends DomainObject<RocketShip> implements RocketShip {
        public static schema = schema;
      }
      it('should not throw error if when valid', () => {
        const ship = new RocketShip({
          serialNumber: uuid(),
          fuelQuantity: 9001,
          passengers: 21,
        });
        expect(ship).toBeInstanceOf(RocketShip); // sanity check
      });
      it('should throw a helpful error when does not pass schema', () => {
        try {
          // eslint-disable-next-line no-new
          new RocketShip({
            serialNumber: '__SOME_UUID__',
            fuelQuantity: 9001,
            passengers: 50,
          });
          throw new Error('should not reach here');
        } catch (error) {
          if (!(error instanceof Error)) throw error;
          expect(error).toBeInstanceOf(HelpfulZodValidationError);
          expect(error.message).toMatchSnapshot();
        }
      });
    });

    describe('hydration', () => {
      it('should hydrate nested domain objects', () => {
        // define the plant pot
        interface PlantPot {
          diameterInInches: number;
        }
        class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

        // define the plant
        interface Plant {
          pot: PlantPot;
          lastWatered: string;
        }
        class Plant extends DomainObject<Plant> implements Plant {
          public static nested = { pot: PlantPot };
        }

        // now show that we hydrate the pot
        const plant = new Plant({
          pot: { diameterInInches: 7 },
          lastWatered: 'monday',
        });
        expect(plant.pot).toBeInstanceOf(PlantPot);
      });
      it('should hydrate nested array of domain objects', () => {
        // define the plant owners
        interface PlantOwner {
          name: string;
        }
        class PlantOwner
          extends DomainObject<PlantOwner>
          implements PlantOwner {}

        // define the plant
        interface Plant {
          owners: PlantOwner[];
          lastWatered: string;
        }
        class Plant extends DomainObject<Plant> implements Plant {
          public static nested = { owners: PlantOwner };
        }

        // now show that we hydrate the pot
        const plant = new Plant({
          owners: [{ name: 'bob' }],
          lastWatered: 'monday',
        });
        plant.owners.forEach((owner) =>
          expect(owner).toBeInstanceOf(PlantOwner),
        );
      });
      it('should not hydrate nullable nested domain objects when null', () => {
        // define the plant owners
        interface PlantOwner {
          name: string;
        }
        class PlantOwner
          extends DomainObject<PlantOwner>
          implements PlantOwner {}

        // define the plant
        interface Plant {
          owners: PlantOwner[] | null;
          lastWatered: string;
        }
        class Plant extends DomainObject<Plant> implements Plant {
          public static nested = { owners: PlantOwner };
        }

        // now show that we hydrate the pot
        const plant = new Plant({ owners: null, lastWatered: 'monday' });
        expect(plant.owners).toEqual(null); // should still be null - since should not have instantiated
      });
      it('should hydrate nested domain objects correctly when given choice of different options', () => {
        // define the plant pot
        interface PlantPot {
          diameterInInches: number;
        }
        class PlantPot extends DomainObject<PlantPot> implements PlantPot {}

        // define the plant bed
        interface PlantBed {
          location: string;
        }
        class PlantBed extends DomainObject<PlantBed> implements PlantBed {}

        // define the plant
        interface Plant {
          plantedIn: PlantPot | PlantBed;
          lastWatered: string;
        }
        class Plant extends DomainObject<Plant> implements Plant {
          public static nested = { plantedIn: [PlantPot, PlantBed] };
        }

        // now show that we hydrate the pot correctly
        const plant = new Plant({
          plantedIn: { _dobj: 'PlantPot', diameterInInches: 7 } as PlantPot,
          lastWatered: 'monday',
        });
        expect(plant.plantedIn).toBeInstanceOf(PlantPot);
      });
    });
  });

  describe('.build', () => {
    describe('.clone', () => {
      it('should be possible to clone an instance ergonomically', () => {
        interface RocketShip {
          serialNumber: string;
          fuelQuantity: number;
          passengers: number;
        }
        class RocketShip
          extends DomainObject<RocketShip>
          implements RocketShip {}
        const ship = RocketShip.build({
          serialNumber: uuid(),
          fuelQuantity: 9001,
          passengers: 21,
        });
        expect(ship).toBeInstanceOf(RocketShip); // sanity check

        // clone it
        const shipB = ship.clone();
        expect(shipB.fuelQuantity).toEqual(9001);
        const shipC = ship.clone({ fuelQuantity: 821 });
        expect(shipC.fuelQuantity).toEqual(821);
      });
    });
  });
});
