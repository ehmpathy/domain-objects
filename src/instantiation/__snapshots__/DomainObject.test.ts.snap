// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DomainObject validation Joi schema should throw a helpful error when does not pass joi schema 1`] = `
"Errors on 1 properties were found while validating properties for domain object RocketShip.:
[
  {
    "message": "\\"serialNumber\\" must be a valid GUID",
    "path": "serialNumber",
    "type": "string.guid"
  }
]

Props Provided:
{
  "serialNumber": "__SOME_UUID__",
  "fuelQuantity": 9001,
  "passengers": 50
}"
`;

exports[`DomainObject validation Yup schema should throw a helpful error when does not pass schema 1`] = `
"Errors were found while validating properties for domain object RocketShip.:
[
  "passengers must be less than or equal to 42"
]

Props Provided:
{
  "serialNumber": "__SOME_UUID__",
  "fuelQuantity": 9001,
  "passengers": 50
}"
`;
